File: node-sidecar-test/test/ejecucion.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';
import * as C from '../src/constantes.js';
import { ClienteDocker } from '../src/cliente-docker.js';
import { Ejecucion } from '../src/ejecucion.js';
import { DaemonDePrueba, frame, type Guion } from './ayudas.js';

/** Secciones 5, 7 y 8 contra un daemon de mentira: la secuencia completa sin Docker real. */

const ID = '3f2b1c4d-0000-4000-8000-000000000001';

async function conDaemon<T>(guion: Guion, cuerpo: (d: DaemonDePrueba, e: Ejecucion) => Promise<T>): Promise<T> {
  const daemon = new DaemonDePrueba(guion);
  await daemon.escuchar();
  try {
    return await cuerpo(daemon, new Ejecucion(new ClienteDocker(daemon.ruta)));
  } finally {
    await daemon.cerrar();
  }
}

/** El stdout que emitiria el entrypoint: el reporte entre los marcadores con el nonce recibido. */
function salidaConReporte(stdinRecibido: Buffer, reporte: string): Buffer {
  const nonce = stdinRecibido.toString('ascii').split('\n')[0] ?? '';
  return frame(1, `hola\n---SANDBOX-${nonce}-INICIO---\n${reporte}\n---SANDBOX-${nonce}-FIN---\n`);
}

test('la secuencia de la seccion 5 corre en orden y borra el contenedor', async () => {
  await conDaemon({ logs: frame(1, 'sin reporte\n') }, async (daemon, ejecucion) => {
    const salida = await ejecucion.ejecutar(ID, Buffer.from('tar de prueba'));

    assert.equal(salida.resultado, 'COMPLETADA');
    assert.equal(salida.exitCode, 0);
    assert.equal(salida.stdout, 'sin reporte\n');
    assert.equal(salida.reporteAusente, true);
    assert.equal(salida.reporte, null);

    const pasos = daemon.llamadas.map((l) => l.replace(/\?.*/, '').replace('contenedor-de-prueba', 'C'));
    assert.deepEqual(pasos, [
      'POST /v1.43/containers/create',      // 1
      'attach /v1.43/containers/C/attach',  // 2 — R5.1: antes del start
      'POST /v1.43/containers/C/start',     // 3
      'POST /v1.43/containers/C/wait',      // 5
      'GET /v1.43/containers/C/json',       // 5b — R5.10
      'GET /v1.43/containers/C/logs',       // 6 — R5.4: antes del delete
      'DELETE /v1.43/containers/C',         // 7 — R5.3
    ]);
    assert.equal(daemon.borrados.length, 1);
    assert.equal(ejecucion.enVuelo.size, 0);
  });
});

test('r5.0 todas las rutas llevan la version de la api explicita en el path', async () => {
  await conDaemon({}, async (daemon, ejecucion) => {
    await ejecucion.ejecutar(ID, Buffer.alloc(0));
    for (const llamada of daemon.llamadas) {
      assert.ok(llamada.includes('/v1.43/'), `sin version en el path: ${llamada}`);
    }
  });
});

test('r7.2 stdin lleva el nonce como primera linea y despues el tar', async () => {
  await conDaemon({}, async (daemon, ejecucion) => {
    const tar = Buffer.from('CONTENIDO-DEL-TAR');
    await ejecucion.ejecutar(ID, tar);

    const salto = daemon.stdinRecibido.indexOf(0x0A);
    const nonce = daemon.stdinRecibido.subarray(0, salto).toString('ascii');
    assert.match(nonce, /^[0-9a-f]{32}$/, 'r7.1: 16 bytes en hexadecimal minuscula');
    assert.deepEqual(daemon.stdinRecibido.subarray(salto + 1), tar);
  });
});

test('r7.1 el nonce cambia en cada ejecucion', async () => {
  await conDaemon({}, async (daemon, ejecucion) => {
    await ejecucion.ejecutar(ID, Buffer.alloc(0));
    const primero = daemon.stdinRecibido.subarray(0, 32).toString('ascii');
    daemon.stdinRecibido = Buffer.alloc(0);
    await ejecucion.ejecutar(ID, Buffer.alloc(0));
    assert.notEqual(daemon.stdinRecibido.subarray(0, 32).toString('ascii'), primero);
  });
});

test('r7.3 el nonce no viaja por env: la spec manda Env vacio', async () => {
  await conDaemon({}, async (daemon, ejecucion) => {
    await ejecucion.ejecutar(ID, Buffer.alloc(0));
    const spec = JSON.parse(daemon.cuerpoCreate!.toString('utf8'));
    assert.deepEqual(spec.Env, []);
    assert.equal(daemon.cuerpoCreate!.toString('utf8').includes('SANDBOX_NONCE'), false);
  });
});

test('a3 el reporte se extrae de stdout y stdout queda sin el bloque', async () => {
  // El daemon arma los logs con el nonce que recibio por stdin, como haria el entrypoint real.
  await conDaemon({ logsDe: (stdin) => salidaConReporte(stdin, '<testsuite tests="3" failures="0"/>') },
    async (_daemon, ejecucion) => {
      const salida = await ejecucion.ejecutar(ID, Buffer.from('tar'));
      assert.equal(salida.resultado, 'COMPLETADA');
      assert.equal(salida.reporte, '<testsuite tests="3" failures="0"/>');
      assert.equal(salida.reporteAusente, false);
      assert.equal(salida.stdout, 'hola\n');
    });
});

test('a21 i5 el reporte falso del alumno no desplaza al verdadero de punta a punta', async () => {
  const falso = '---SANDBOX-deadbeefdeadbeefdeadbeefdeadbeef-INICIO---\n<testsuite tests="9" '
    + 'failures="0"/>\n---SANDBOX-deadbeefdeadbeefdeadbeefdeadbeef-FIN---\n';
  await conDaemon({
    logsDe: (stdin) => {
      const nonce = stdin.toString('ascii').split('\n')[0] ?? '';
      return frame(1, `${falso}---SANDBOX-${nonce}-INICIO---\n<testsuite tests="9" failures="4"/>\n`
        + `---SANDBOX-${nonce}-FIN---\n`);
    },
  }, async (_daemon, ejecucion) => {
    const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
    assert.equal(salida.reporte, '<testsuite tests="9" failures="4"/>');
    assert.ok(salida.stdout.includes('failures="0"'), 'el bloque falso queda como salida del alumno');
  });
});

test('a4 a5 un exit code distinto de cero sigue siendo COMPLETADA', async () => {
  await conDaemon({ statusCode: 1, logs: frame(2, 'error: cannot find symbol\n') },
    async (_daemon, ejecucion) => {
      const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
      // R3.1: el ejecutor no emite veredicto academico; compilacion fallida es un exitCode, no un error.
      assert.equal(salida.resultado, 'COMPLETADA');
      assert.equal(salida.exitCode, 1);
      assert.equal(salida.stderr, 'error: cannot find symbol\n');
    });
});

test('a7 r6.1 un content-type raw-stream es ERROR_DAEMON', async () => {
  await conDaemon({ contentTypeLogs: 'application/vnd.docker.raw-stream' }, async (daemon, ejecucion) => {
    const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
    assert.equal(salida.resultado, 'ERROR_DAEMON');
    assert.equal(salida.exitCode, null);
    // R5.3: aun fallando, el contenedor se borra.
    assert.equal(daemon.borrados.length, 1);
  });
});

test('r5.3 un create que falla no deja contenedor ni rompe el proceso', async () => {
  await conDaemon({ codigoCreate: 500 }, async (daemon, ejecucion) => {
    const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
    assert.equal(salida.resultado, 'ERROR_DAEMON');
    assert.equal(daemon.borrados.length, 0, 'no hay nada que borrar si create fallo');
    assert.equal(ejecucion.enVuelo.size, 0);
  });
});

test('r3.2 el mensaje de error no filtra la ruta del socket ni el cuerpo del daemon', async () => {
  await conDaemon({ codigoCreate: 500 }, async (daemon, ejecucion) => {
    const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
    const json = JSON.stringify(salida);
    assert.equal(json.includes(daemon.ruta), false);
    assert.equal(json.includes('roto'), false);
  });
});

test('a31 c2 oomKilled viene del inspect, no del exit code', async () => {
  await conDaemon({ oomKilled: true, statusCode: 3 }, async (_daemon, ejecucion) => {
    const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
    assert.equal(salida.oomKilled, true);
    assert.equal(salida.exitCode, 3);
    assert.equal(salida.resultado, 'COMPLETADA');
  });
});

test('a31 r5.10 un inspect que falla no aborta la ejecucion', async () => {
  await conDaemon({ inspectFalla: true, statusCode: 7, logs: frame(1, 'salida\n') },
    async (daemon, ejecucion) => {
      const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
      // El resultado queda intacto; solo se pierde el dato de diagnostico.
      assert.equal(salida.oomKilled, false);
      assert.equal(salida.resultado, 'COMPLETADA');
      assert.equal(salida.exitCode, 7);
      assert.equal(salida.stdout, 'salida\n');
      assert.equal(daemon.borrados.length, 1);
    });
});

test('r3.4 un exit code 137 sin OOMKilled no se reinterpreta', async () => {
  await conDaemon({ statusCode: 137, oomKilled: false }, async (_daemon, ejecucion) => {
    const salida = await ejecucion.ejecutar(ID, Buffer.alloc(0));
    assert.equal(salida.exitCode, 137);
    assert.equal(salida.oomKilled, false, 'no se infiere del exitCode');
  });
});

// ---------------------------------------------------------------- los que consumen el reloj entero

test('r8.2 r5.10 al vencer el reloj hay kill, logs, inspect y delete, y el resultado es TIMEOUT',
  { timeout: 120_000 }, async () => {
    await conDaemon({ demoraWaitMs: C.TIMEOUT_EJECUCION_MS * 3, oomKilled: true, logs: frame(1, 'parcial\n') },
      async (daemon, ejecucion) => {
        const salida = await ejecucion.ejecutar(ID, Buffer.from('tar'));

        assert.equal(salida.resultado, 'TIMEOUT');
        assert.equal(salida.exitCode, null);
        // La salida parcial se devuelve igual: sirve para diagnosticar.
        assert.equal(salida.stdout, 'parcial\n');
        // R5.10: el paso 5b corre tambien en TIMEOUT.
        assert.equal(salida.oomKilled, true);

        const pasos = daemon.llamadas.map((l) => l.split(' ')[1]!.replace(/\?.*/, ''));
        assert.ok(pasos.some((p) => p.endsWith('/kill')));
        assert.ok(pasos.some((p) => p.endsWith('/logs')));
        assert.equal(daemon.borrados.length, 1);
      });
  });

test('a33 c3 r8.5 una escritura que no avanza no retiene el cupo de concurrencia',
  { timeout: 180_000 }, async () => {
    // El daemon adjunta el socket y no lee nunca: el buffer se llena y el paso 4 queda bloqueado.
    await conDaemon({ consumeStdin: false, demoraWaitMs: C.TIMEOUT_EJECUCION_MS * 3 },
      async (daemon, ejecucion) => {
        const comienzo = Date.now();
        const salida = await ejecucion.ejecutar(ID, Buffer.alloc(C.MAX_BUNDLE_BYTES, 0x41));
        const transcurrido = Date.now() - comienzo;

        // Sin el tope de R8.5 esto no termina nunca y el cupo no se libera jamas.
        assert.ok(transcurrido < C.TIMEOUT_EJECUCION_MS * 2.5,
          `la ejecucion tardo ${transcurrido} ms: la escritura no abandono`);
        assert.equal(salida.resultado, 'TIMEOUT');
        // El cupo se libera: el contenedor se borro y no queda nada en vuelo.
        assert.equal(ejecucion.enVuelo.size, 0);
        assert.equal(daemon.borrados.length, 1);
      });
  });

File: java sidecar test/src/test/java/sandbox/ejecutor/EjecucionTest.java
package sandbox.ejecutor;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

import static org.junit.jupiter.api.Assertions.*;

/** Seccion 5 completa contra un daemon de mentira: orden de llamadas, stdin, logs y limpieza. */
class EjecucionTest {

    private static final String NONCE_REGEX = "[0-9a-f]{32}";

    private Path directorio;
    private DaemonDePrueba daemon;
    private ClienteDocker cliente;
    private Ejecucion ejecucion;
    private ExecutorService hilos;
    private ScheduledExecutorService reloj;

    @BeforeEach
    void levantar() throws Exception {
        directorio = Files.createTempDirectory("dk");
        daemon = new DaemonDePrueba(directorio.resolve("d.sock"));
        reloj = Executors.newScheduledThreadPool(2);
        hilos = Executors.newVirtualThreadPerTaskExecutor();
        cliente = new ClienteDocker(daemon.rutaSocket(), reloj);
        ejecucion = new Ejecucion(cliente, hilos, reloj);
    }

    @AfterEach
    void bajar() throws Exception {
        daemon.close();
        hilos.shutdownNow();
        reloj.shutdownNow();
        try (var f = Files.walk(directorio)) {
            f.sorted(java.util.Comparator.reverseOrder()).forEach(p -> p.toFile().delete());
        }
    }

    private static byte[] concatenar(byte[]... partes) {
        ByteArrayOutputStream o = new ByteArrayOutputStream();
        for (byte[] p : partes) o.writeBytes(p);
        return o.toByteArray();
    }

    private static String id() { return UUID.randomUUID().toString(); }

    /** El camino feliz: los siete pasos, en orden, con el reporte separado del stdout. */
    @Test
    @Timeout(30)
    void caminoFeliz() {
        // El nonce lo elige el ejecutor, asi que el "entrypoint" de mentira no puede saberlo de antemano.
        // Se resuelve en dos etapas: primero corremos para capturar el nonce que entro por stdin.
        daemon.logs = concatenar(
                DaemonDePrueba.frame(1, "compilando\n"),
                DaemonDePrueba.frame(2, "un aviso\n"));

        String ejecucionId = id();
        var salida = ejecucion.ejecutar(ejecucionId, "TAR".getBytes(StandardCharsets.UTF_8));

        assertEquals(Ejecucion.Estado.COMPLETADA, salida.resultado());
        assertEquals(0, salida.exitCode());
        assertEquals("compilando\n", salida.stdout());
        assertEquals("un aviso\n", salida.stderr());
        assertTrue(salida.reporteAusente());
        assertNull(salida.reporte());
        assertFalse(salida.salidaTruncada());

        // R5.1 y R5.4: attach antes de start, logs antes de delete.
        int attach = indiceDe("/attach");
        int start = indiceDe("/start");
        int logs = indiceDe("/logs");
        int borrado = indiceDe("?force=1");
        assertTrue(attach < start, "el attach tiene que ir antes del start");
        assertTrue(logs < borrado, "los logs se leen antes del delete");
        assertTrue(indiceDe("/wait") < logs);
    }

    /** R7.1 y R7.2: el nonce es la primera linea de stdin, y despues van los bytes del tar. */
    @Test
    @Timeout(30)
    void r72_stdinLlevaNonceYDespuesElTar() {
        byte[] tar = "contenido del tar, opaco".getBytes(StandardCharsets.UTF_8);
        ejecucion.ejecutar(id(), tar);

        byte[] recibido = daemon.stdinRecibido;
        assertNotNull(recibido, "no llego nada por el socket adjunto");
        String texto = new String(recibido, StandardCharsets.UTF_8);
        int salto = texto.indexOf('\n');
        assertTrue(salto > 0, "el nonce tiene que terminar en salto de linea");

        String nonce = texto.substring(0, salto);
        assertTrue(nonce.matches(NONCE_REGEX), "nonce de 32 hex minusculas, era: " + nonce.length() + " chars");
        assertArrayEquals(tar, java.util.Arrays.copyOfRange(recibido, salto + 1, recibido.length),
                "despues del nonce van los bytes del tar, sin tocar");
        assertTrue(daemon.stdinCerrado, "el socket adjunto se cierra entero");
    }

    /** I7: el ejecutor no desempaqueta ni interpreta el tar; lo reenvia byte a byte. */
    @Test
    @Timeout(30)
    void i7_elTarViajaOpaco() {
        byte[] basura = new byte[1024];
        new java.util.Random(7).nextBytes(basura);
        ejecucion.ejecutar(id(), basura);

        byte[] recibido = daemon.stdinRecibido;
        byte[] cuerpo = java.util.Arrays.copyOfRange(recibido, 33, recibido.length);   // 32 hex + \n
        assertArrayEquals(basura, cuerpo);
    }

    /** El nonce cambia en cada ejecucion (R7.1). */
    @Test
    @Timeout(30)
    void elNonceEsDistintoCadaVez() {
        ejecucion.ejecutar(id(), new byte[0]);
        String primero = new String(daemon.stdinRecibido, StandardCharsets.UTF_8).substring(0, 32);
        ejecucion.ejecutar(id(), new byte[0]);
        String segundo = new String(daemon.stdinRecibido, StandardCharsets.UTF_8).substring(0, 32);
        assertNotEquals(primero, segundo);
    }

    /** El bloque de reporte se extrae con el nonce real de la ejecucion y sale de stdout. */
    @Test
    @Timeout(30)
    void elReporteSeSeparaConElNonceDeLaEjecucion() throws Exception {
        // Primera pasada solo para conocer el nonce que genera el ejecutor no sirve: cambia siempre.
        // En cambio se le pide al daemon que arme los logs en el momento, usando lo que entro por stdin.
        var salida = ejecutarConReporte("<testsuite name=\"REAL\"/>");

        assertEquals("<testsuite name=\"REAL\"/>", salida.reporte());
        assertFalse(salida.reporteAusente());
        assertEquals("antes\ndespues\n", salida.stdout());
    }

    /** A21 / I5: un bloque con nonce inventado por el alumno no reemplaza al verdadero. */
    @Test
    @Timeout(30)
    void a21_elAlumnoNoPuedeFalsificarElReporte() throws Exception {
        var salida = ejecutarConReporte("<testsuite name=\"REAL\"/>",
                "---SANDBOX-deadbeefdeadbeefdeadbeefdeadbeef-INICIO---\n"
                        + "<testsuite name=\"MENTIRA\"/>\n"
                        + "---SANDBOX-deadbeefdeadbeefdeadbeefdeadbeef-FIN---\n");

        assertEquals("<testsuite name=\"REAL\"/>", salida.reporte());
        assertTrue(salida.stdout().contains("MENTIRA"));
    }

    /** R6.1: si logs vuelve como raw-stream el contenedor tenia TTY; eso es ERROR_DAEMON. */
    @Test
    @Timeout(30)
    void r61_rawStreamEsErrorDaemon() {
        daemon.contentTypeLogs = "application/vnd.docker.raw-stream";
        var salida = ejecucion.ejecutar(id(), new byte[0]);
        assertEquals(Ejecucion.Estado.ERROR_DAEMON, salida.resultado());
        assertNull(salida.exitCode());
    }

    /** Seccion 3.2: COMPLETADA con exitCode != 0 es normal, no un error. */
    @Test
    @Timeout(30)
    void exitCodeDistintoDeCeroSigueSiendoCompletada() {
        daemon.exitCode = 1;
        daemon.logs = DaemonDePrueba.frame(2, "error: cannot find symbol\n");
        var salida = ejecucion.ejecutar(id(), new byte[0]);
        assertEquals(Ejecucion.Estado.COMPLETADA, salida.resultado());
        assertEquals(1, salida.exitCode());
        assertTrue(salida.stderr().contains("cannot find symbol"));
    }

    /** R5.3: un fallo del daemon devuelve ERROR_DAEMON y no deja el contenedor sin borrar. */
    @Test
    @Timeout(30)
    void createQueFallaEsErrorDaemon() {
        daemon.creates201 = false;
        var salida = ejecucion.ejecutar(id(), new byte[0]);
        assertEquals(Ejecucion.Estado.ERROR_DAEMON, salida.resultado());
        assertTrue(ejecucion.contenedoresEnVuelo().isEmpty());
    }

    /** R5.3: el delete corre siempre, y al terminar no queda nada anotado como en vuelo. */
    @Test
    @Timeout(30)
    void siempreSeBorraElContenedor() {
        ejecucion.ejecutar(id(), new byte[0]);
        assertTrue(daemon.llamadas.stream().anyMatch(l -> l.contains("?force=1&v=1")),
                "el delete tiene que ir con force y v");
        assertTrue(ejecucion.contenedoresEnVuelo().isEmpty());
    }

    /**
     * R8.2: vencido el reloj de ejecucion se hace kill, se vuelve a esperar, se leen igual los logs
     * parciales y se borra. El resultado es TIMEOUT con exitCode null.
     */
    @Test
    @Timeout(120)
    void r82_timeoutMataYDevuelveLaSalidaParcial() {
        daemon.demoraWaitMs = Constantes.TIMEOUT_EJECUCION_MS * 3;
        daemon.logs = DaemonDePrueba.frame(1, "arranque y me colgue\n");

        var salida = ejecucion.ejecutar(id(), new byte[0]);

        assertEquals(Ejecucion.Estado.TIMEOUT, salida.resultado());
        assertNull(salida.exitCode());
        assertEquals("arranque y me colgue\n", salida.stdout(), "la salida parcial se devuelve igual");
        assertTrue(daemon.llamadas.stream().anyMatch(l -> l.contains("/kill")));
        assertTrue(daemon.llamadas.stream().anyMatch(l -> l.contains("?force=1")));
        assertTrue(salida.duracionMs() >= Constantes.TIMEOUT_EJECUCION_MS);
    }

    /** R5.0: la version de la API va explicita en el path, nunca el default del daemon. */
    @Test
    @Timeout(30)
    void r50_todasLasRutasLlevanLaVersion() {
        ejecucion.ejecutar(id(), new byte[0]);
        for (String llamada : daemon.llamadas) {
            assertTrue(llamada.startsWith("/" + Constantes.VERSION_API_DOCKER + "/"),
                    "ruta sin version: " + llamada);
        }
    }

    /** A1 en vivo: el cuerpo que recibe el daemon es exactamente el golden. */
    @Test
    @Timeout(30)
    void elCuerpoDeCreateEsElGolden() throws Exception {
        String ejecucionId = id();
        ejecucion.ejecutar(ejecucionId, new byte[0]);

        String ruta = daemon.llamadas.stream().filter(l -> l.contains("/create")).findFirst().orElseThrow();
        assertTrue(ruta.endsWith("?name=sandbox-" + ejecucionId));

        byte[] enviado = daemon.cuerpos.get(ruta);
        assertArrayEquals(Spec.crear(ejecucionId), enviado);
    }

    /** R5.7: la respuesta de logs viene chunked y el cliente la reconstruye entera. */
    @Test
    @Timeout(30)
    void r57_logsChunkedSeReconstruyeEntero() {
        String largo = "x".repeat(5000);
        daemon.logs = DaemonDePrueba.frame(1, largo);
        var salida = ejecucion.ejecutar(id(), new byte[0]);
        assertEquals(largo, salida.stdout());
    }

    @Test
    void pingContraElDaemon() {
        assertTrue(cliente.ping());
    }

    // ------------------------------------------------------------------ seccion 13.6

    /** A31 (C2): el oomKilled sale del inspect del paso 5b. */
    @Test
    @Timeout(30)
    void a31_oomKilledVieneDelInspect() {
        daemon.oomKilled = true;
        // R3.4: con la JVM bien configurada el OOM llega como exitCode 3 y OOMKilled false, o al
        // reves. Por eso el dato no se infiere del exitCode: aca el exit es 0 y el oom es true.
        daemon.exitCode = 0;

        var salida = ejecucion.ejecutar(id(), new byte[0]);

        assertTrue(salida.oomKilled());
        assertEquals(Ejecucion.Estado.COMPLETADA, salida.resultado());
        assertEquals(0, salida.exitCode());
        // R5.10: el inspect va despues del wait y antes del delete.
        assertTrue(indiceDe("/wait") < indiceDe("/contenedor-de-prueba/json"));
        assertTrue(indiceDe("/contenedor-de-prueba/json") < indiceDe("?force=1"));
    }

    /** A31 (C2): si el inspect falla, se devuelve oomKilled=false y el resto del resultado intacto. */
    @Test
    @Timeout(30)
    void a31_elInspectQueFallaNoAbortaLaEjecucion() {
        daemon.inspectFalla = true;
        daemon.exitCode = 7;
        daemon.logs = DaemonDePrueba.frame(1, "salida normal\n");

        var salida = ejecucion.ejecutar(id(), new byte[0]);

        assertFalse(salida.oomKilled());
        assertEquals(Ejecucion.Estado.COMPLETADA, salida.resultado(), "un dato de diagnostico no puede tirar el resultado");
        assertEquals(7, salida.exitCode());
        assertEquals("salida normal\n", salida.stdout());
    }

    /** R5.10: en TIMEOUT el paso 5b se ejecuta igual. */
    @Test
    @Timeout(180)
    void r510_elInspectTambienCorreEnTimeout() {
        daemon.demoraWaitMs = Constantes.TIMEOUT_EJECUCION_MS * 2;
        daemon.oomKilled = true;

        var salida = ejecucion.ejecutar(id(), new byte[0]);

        assertEquals(Ejecucion.Estado.TIMEOUT, salida.resultado());
        assertTrue(salida.oomKilled(), "un contenedor matado por memoria que ademas llego al reloj es un caso real");
    }

    /**
     * A33 (C3): un contenedor que arranca y no consume stdin. Sin el tope de R8.5 la escritura del
     * paso 4 queda bloqueada para siempre y el cupo de concurrencia no se libera nunca; el bug es
     * invisible porque la suite queda verde y el ejecutor se traba en produccion.
     */
    @Test
    @Timeout(180)
    void a33_laEscrituraQueNoAvanzaNoRetieneElCupo() {
        daemon.consumeStdin = false;
        byte[] bundleGrande = new byte[Constantes.MAX_BUNDLE_BYTES];   // muy por encima del buffer del pipe

        long comienzo = System.nanoTime();
        var salida = ejecucion.ejecutar(id(), bundleGrande);
        long duracionMs = (System.nanoTime() - comienzo) / 1_000_000;

        assertEquals(Ejecucion.Estado.TIMEOUT, salida.resultado());
        assertNull(salida.exitCode());
        // Lo que importa: la llamada RETORNA. Sin R8.5 nunca lo haria.
        assertTrue(duracionMs < Constantes.TIMEOUT_EJECUCION_MS * 2,
                "la escritura no se abandono a tiempo: " + duracionMs + " ms");
        // Y el contenedor queda borrado, asi que el cupo se libera limpio.
        assertTrue(ejecucion.contenedoresEnVuelo().isEmpty());
    }

    // ------------------------------------------------------------------ apoyo

    private int indiceDe(String fragmento) {
        for (int i = 0; i < daemon.llamadas.size(); i++) {
            if (daemon.llamadas.get(i).contains(fragmento)) return i;
        }
        return fail("nunca se llamo a " + fragmento);
    }

    /**
     * Corre una ejecucion en la que el daemon arma los logs con el nonce real, leyendolo del stdin
     * que acaba de recibir: es lo que hace el entrypoint de la imagen en produccion.
     */
    private Ejecucion.Salida ejecutarConReporte(String contenido, String... ruidoPrevio) throws Exception {
        DaemonDePrueba d = daemon;
        Thread armador = new Thread(() -> {
            while (d.stdinRecibido == null) Thread.onSpinWait();
            String nonce = new String(d.stdinRecibido, StandardCharsets.UTF_8).substring(0, 32);
            StringBuilder salida = new StringBuilder("antes\n");
            for (String r : ruidoPrevio) salida.append(r);
            salida.append("---SANDBOX-").append(nonce).append("-INICIO---\n")
                  .append(contenido).append('\n')
                  .append("---SANDBOX-").append(nonce).append("-FIN---\n")
                  .append("despues\n");
            d.logs = DaemonDePrueba.frame(1, salida.toString());
        });
        armador.setDaemon(true);
        armador.start();

        daemon.demoraWaitMs = 300;   // le da tiempo al armador antes de que se pidan los logs
        Ejecucion.Salida salida = ejecucion.ejecutar(id(), new byte[0]);
        armador.join(1000);
        return salida;
    }
}
